listenerutil/handler.go

```diff
@@ -0,0 +1,109 @@
+package listenerutil
+
+import (
+	"net/http"
+)
+
+type ResponseWriter struct {
+	wrapped http.ResponseWriter
+	// headers contain a map of response code to header map such that
+	// headers[status][header name] = header value
+	// this map also contains values for hundred-level values in the format 1: "1xx", 2: "2xx", etc
+	// defaults are set to 0
+	headers       map[int]map[string][]string
+	headerWritten bool
+}
+
+func (w *ResponseWriter) WriteHeader(statusCode int) {
+	w.headerWritten = true
+	w.setCustomResponseHeaders(statusCode)
+	w.wrapped.WriteHeader(statusCode)
+}
+
+func (w *ResponseWriter) Header() http.Header {
+	return w.wrapped.Header()
+}
+
+func (w *ResponseWriter) Write(data []byte) (int, error) {
+	// The default behavior of http.ResponseWriter.Write is such that if WriteHeader has not
+	// yet been called, it calls it with the below line. We will copy that logic so that our
+	// WriteHeader function is called rather than http.ResponseWriter.WriteHeader
+	if !w.headerWritten {
+		w.WriteHeader(http.StatusOK)
+	}
+	return w.wrapped.Write(data)
+}
+
+func (w *ResponseWriter) setCustomResponseHeaders(statusCode int) {
+	sch := w.headers
+	if sch == nil {
+		return
+	}
+
+	// Check the validity of the status code
+	if statusCode >= 600 || statusCode < 100 {
+		return
+	}
+
+	// Setter function to set headers
+	setter := func(headerMap map[string][]string) {
+		for header, values := range headerMap {
+			w.Header().Del(header)
+			for _, value := range values {
+				w.Header().Add(header, value)
+			}
+		}
+	}
+
+	// Setting the default headers first
+	if val, ok := sch[0]; ok {
+		setter(val)
+	}
+
+	// Then setting the generic hundred-level headers
+	// Note: integer division always rounds down, so 499/100 = 4
+	if val, ok := sch[statusCode/100]; ok {
+		setter(val)
+	}
+
+	// Finally setting the status-specific headers
+	if val, ok := sch[statusCode]; ok {
+		setter(val)
+	}
+}
+
+type uiRequestFunc func(*http.Request) bool
+
+// WrapCustomHeadersHandler wraps the handler to pass a custom ResponseWriter struct to all
+// later wrappers and handlers to assign custom headers by status code. This wrapper must
+// be the outermost wrapper to function correctly.
+func WrapCustomHeadersHandler(h http.Handler, config *ListenerConfig, isUiRequest uiRequestFunc) http.Handler {
+	// TODO: maybe we should perform some preparsing here on the headers? check for duplicates,
+	// headers that aren't allowed, etc.
```

Comment:
Do you want to remove this comment? I would expect this sort of parsing to happen in the parser, not here.